tests/all/piped_tests.rs

```diff
@@ -0,0 +1,72 @@
+#![cfg(not(miri))]
+
+use super::cli_tests::get_wasmtime_command;
+use anyhow::Result;
+use std::process::{ChildStdout, Command, Stdio};
+
+fn producer_command(component_path: &str) -> Result<Command> {
+    let mut cmd = get_wasmtime_command()?;
+    cmd.arg("run")
+        .arg("-Wcomponent-model")
+        .arg("--env")
+        .arg("PIPED_SIDE=PRODUCER")
+        .arg(component_path)
+        .stdout(Stdio::piped());
+    Ok(cmd)
+}
+
+fn consumer_command(component_path: &str, input: ChildStdout) -> Result<Command> {
+    let mut cmd = get_wasmtime_command()?;
+    cmd.arg("run")
+        .arg("-Wcomponent-model")
+        .arg("--env")
+        .arg("PIPED_SIDE=CONSUMER")
+        .arg(component_path)
+        .stdin(input);
+    Ok(cmd)
+}
+
+pub fn run_wasmtime_piped(component_path: &str) -> Result<()> {
+    let mut producer = producer_command(component_path)?.spawn()?;
+    let stdout = producer.stdout.take().unwrap();
+    let mut consumer = consumer_command(component_path, stdout)?.spawn()?;
+
+    let producer = producer.wait()?;
+    if !producer.success() {
+        // make sure the consumer gets killed off
+        if consumer.try_wait().is_err() {
+            consumer.kill().expect("Failed to kill consumer");
```

Comment:
rather than kill the consumer, can we tell each the producer and consumer how many bytes or messages to send/expect total, so they can both exit gracefully once those conditions are met? this will also let us assert, in the consumer, that after everything expected is consumed, the pollable does not become ready "ever again" (poll with a 250ms timeout and call it good)

Reply:
If the producer panics, we'll still want to kill the consumer. This was more about ensuring that we cleanup failures properly than anything else.

Reply:
Additionally, consumed amount is something that we could compute in the individual piped tests, and I think would be easier to debug as an assertion failure in the consumer side.

Reply:
got it! sounds good.